gadgets/trace_dns/test/unit/trace_dns_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2025 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"flag"
+	"fmt"
+	"log"
+
+	"testing"
+	"time"
+
+	gadgettesting "github.com/inspektor-gadget/inspektor-gadget/gadgets/testing"
+	utilstest "github.com/inspektor-gadget/inspektor-gadget/internal/test"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/gadgetrunner"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/utils"
+	"github.com/miekg/dns"
+)
+
+type ExpectedTraceDnsEvent struct { // copied from integration test for now
+	Timestamp string        `json:"timestamp"`
+	NetNsID   uint64        `json:"netns_id"`
+	Proc      utils.Process `json:"proc"`
+
+	Src utils.L4Endpoint `json:"src"`
+	Dst utils.L4Endpoint `json:"dst"`
+
+	// Raw fields are coming from wasm, test them too
+	ID                 string `json:"id"`
+	Qtype              string `json:"qtype"`
+	QtypeRaw           uint16 `json:"qtype_raw"`
+	PktType            string `json:"pkt_type"`
+	RcodeRaw           uint16 `json:"rcode_raw"`
+	Rcode              string `json:"rcode"`
+	Latency            uint64 `json:"latency_ns_raw"`
+	QrRaw              bool   `json:"qr_raw"`
+	Qr                 string `json:"qr"`
+	Name               string `json:"name"`
+	Addresses          string `json:"addresses"`
+	Truncated          bool   `json:"tc"`
+	RecursionDesired   bool   `json:"rd"`
+	RecursionAvailable bool   `json:"ra"`
+}
+
+type testDef struct {
+	name         string
+	runnerConfig *utilstest.RunnerConfig
+}
+
+func printEvents(t *testing.T, events []ExpectedTraceDnsEvent) {
```

Comment:
nit: you can use fmt.Printf("%+v", event) to print the fields with their name. Another possibility is to marshal and print it as json.